### PR TITLE
WIP: Dataframe API ML preprocessing notebook

### PR DESCRIPTION
**Pull request by PhilippeMoussalli**
- PR that implements a notebook to demonstrate the usage of the beam dataframe API as a preprocessing tool for ML training

WIP:
- [x] **Find a method to implement the one-hot-encoding for encoding categorical variables:** related to ticket [#22268](https://github.com/apache/beam/issues/22268)
- [x]  **Fix bug that returns `ValueError: No producer for ref_PCollection_PCollection_265` when attempting to merge two deferred datasets :** related to ticket [#22267](https://github.com/apache/beam/issues/22267)

### CODE REVIEW

**Comment by PhilippeMoussalli** [comment]
@rezarokni @TheNeuralBit

**Comment by TheNeuralBit** [comment]
I think the "No producer" error is a bug with PCollection pruning in interactive beam: #21430

CC: @KevinGG

**Comment by KevinGG** [comment]
> I think the "No producer" error is a bug with PCollection pruning in interactive beam: #21430
> 
> CC: @KevinGG

Commented in https://github.com/apache/beam/issues/21430, we can disable pruning for dataframe like what we did for TestStream: https://github.com/apache/beam/blob/master/sdks/python/apache_beam/runners/interactive/pipeline_fragment.py#L219

**Comment by TheNeuralBit** [comment]
>  Have only one installation script for Beam with the latest implemented functions in the Dataframe API instead of installing from source

This is just blocked on the 2.41.0 release, right?

**Comment by TheNeuralBit** [comment]
> release

Note if we have to do that to unblock this change it will be blocked until 2.42.0 is out.

**Comment by TheNeuralBit** [comment]
> > I think the "No producer" error is a bug with PCollection pruning in interactive beam: #21430
> > CC: @KevinGG
> 
> Commented in #21430, we can disable pruning for dataframe like what we did for TestStream: https://github.com/apache/beam/blob/master/sdks/python/apache_beam/runners/interactive/pipeline_fragment.py#L219

Note if we have to do that to unblock this change, it will be blocked until 2.42.0 is out.

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (TheNeuralBit) [inline examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:1777]: Should `sample.csv` be renamed `full.csv`? The name sample makes me think it's still a subset of the full dataset.

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (TheNeuralBit) [inline examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:1616]: `str` is an attribute on Series, you need to access just a Series (e.g. `object_class_col = beam_df.object_class`) in order to use `str.get_dummies`. Selecting a single column with `filter` produces a single column DataFrame.

You will also need to configure that column to have a `CategoricalDtype`. This is where you can use `unique` and `ib.collect`, something like:
```
unique_classes = ib.collect(object_class_col.unique())
object_class_col.astype(pd.CategoricalDtype(unique_classes)).str.get_dummies()
```

**Comment by PhilippeMoussalli** [comment]
> > Have only one installation script for Beam with the latest implemented functions in the Dataframe API instead of installing from source
> 
> This is just blocked on the 2.41.0 release, right?

I am aware of the particular release date of 2.41.0. I suppose it will depend if we manage to resolve all the friction points before that release.

**Comment by PhilippeMoussalli** [comment]
> > > I think the "No producer" error is a bug with PCollection pruning in interactive beam: #21430
> > > CC: @KevinGG
> > 
> > 
> > Commented in #21430, we can disable pruning for dataframe like what we did for TestStream: https://github.com/apache/beam/blob/master/sdks/python/apache_beam/runners/interactive/pipeline_fragment.py#L219
> 
> Note if we have to do that to unblock this change, it will be blocked until 2.42.0 is out.

Would it be easier to execute the work-around with  ‘loc.setitem’?    https://github.com/apache/beam/issues/22267

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (PhilippeMoussalli) [reply examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:1777]: good remark ;), indeed I think `full.csv` makes more sense. I'll change it accordingly.

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (PhilippeMoussalli) [reply examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:1616]: I tried this approach and I am still not able to run the `unique()` command event when defining the argument `unique(as_series=True)` as specified [here](https://beam.apache.org/releases/pydoc/current/apache_beam.dataframe.frames.html#:~:text=unique(as_series%3DFalse,%5Bsource%5D). 
(Error: `unique()` is not implemented for deferred Dataframes`).
In any case, I think it would be nicer to wait up on the `get_dummies()` as it will be more intuitive to use for user. I see that you already filed a [ticket](https://github.com/apache/beam/issues/22646) for it and referred to it in the deliverable 3 doc.

**Comment by KevinGG** [comment]
> > > > I think the "No producer" error is a bug with PCollection pruning in interactive beam: #21430
> > > > CC: @KevinGG
> > > 
> > > 
> > > Commented in #21430, we can disable pruning for dataframe like what we did for TestStream: https://github.com/apache/beam/blob/master/sdks/python/apache_beam/runners/interactive/pipeline_fragment.py#L219
> > 
> > 
> > Note if we have to do that to unblock this change, it will be blocked until 2.42.0 is out.
> 
> Would it be easier to execute the work-around with ‘loc.setitem’? #22267

The work-around is applied to a specific typed composite transform. So the difficulty is the same.

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (rezarokni) [inline examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:41]: Describe that as we want to explore the elements within a PCollection we can make use of the ...

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (rezarokni) [inline examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:55]: Remove.

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (rezarokni) [inline examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:102]: Test our code interactively, building out the pipeline as we go

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (rezarokni) [inline examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:673]: can be used, or need to be used?

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (rezarokni) [inline examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:1171]: Maybe a link ?

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (rezarokni) [inline examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:1726]: Can the location be a string variable, will help with readability.

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (TheNeuralBit) [reply examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:1616]: Did you try `unique` on just the series that you are trying to one-hot encode? Note that even `pd.get_dummies`, the user is going to have to create a `CategoricalDtype` somehow, which will likely utilize `unique`, so we need to get that working regardless.

Comment (TheNeuralBit) [reply examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:1616]: To be clear, I'm suggesting you change `object_class_col= beam_df.filter(items=['object_class'])` to `object_class_col= beam_df['object_class']`

When you select the column with `filter` it's creating a DataFrame with a single column, not a series.

Comment (TheNeuralBit) [reply examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:1616]: Also it looks like the pandas `DataFrame` doesn't actually have a `unique` operation. If the Beam Deferred DataFrame is giving "unique() is not implemented for deferred Dataframe" then that's a bug. It should give an `AttributeError` like pandas does.

To be sure, I just checked and we do give the same error as pandas:
```
In [7]: df.unique
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Input In [7], in <cell line: 1>()
----> 1 beam_df.unique

File ~/working_dir/beam/sdks/python/apache_beam/dataframe/frames.py:2484, in DeferredDataFrame.__getattr__(self, name)
   2482   return self[name]
   2483 else:
-> 2484   return object.__getattribute__(self, name)

AttributeError: 'DeferredDataFrame' object has no attribute 'unique'
```

If a user wants the unique values in a `DataFrame` they can use `drop_duplicates`, as suggested in https://stackoverflow.com/questions/43184491/df-unique-on-whole-dataframe-based-on-a-column.

`drop_duplicates` (with `keep='any'`) should work in Beam.

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (PhilippeMoussalli) [reply examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:41]: Good point :)

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (PhilippeMoussalli) [reply examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:673]: woops need to be used indeed :) nice catch!

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (PhilippeMoussalli) [reply examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:1617]: @TheNeuralBit indeed you're correct `unique()` can only be applied to `Series`. I was able to get the unique classes of the categorical column and transform it to a `CategoricalDtype`:

```
object_class_col= beam_df['object_class']

unique_classes = pd.CategoricalDtype(ib.collect(object_class_col.unique(as_series=True)))
```
This returns 

```
CategoricalDtype(categories=['MBA', 'OMB', 'MCA', 'AMO', 'IMB', 'TJN', 'CEN', 'APO',
                  'ATE', 'AST'], ordered=False)
```

I have tried implementing the workaround  you suggested for `str.get_dummies()` but still ran into some issues:

```
object_class_col.astype(unique_classes).str.get_dummies()

WontImplementError: astype(dtype='category') is not supported because the type of the output column depends on the data. Please use pd.CategoricalDtype with explicit categories instead.
For more information see https://s.apache.org/dataframe-non-deferred-columns.
```
I ran into the same issue when trying to covert the `object_class_col` to a categorical variable:

`
object_class_col.astype('category')
`

Am I still missing something?

**Comment by PhilippeMoussalli** [comment]
> > I think the "No producer" error is a bug with PCollection pruning in interactive beam: #21430
> > CC: @KevinGG
> 
> Commented in #21430, we can disable pruning for dataframe like what we did for TestStream: https://github.com/apache/beam/blob/master/sdks/python/apache_beam/runners/interactive/pipeline_fragment.py#L219

Is there any update on this or a potential workaround for merging Deferred dataframes?

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (PhilippeMoussalli) [reply examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:1730]: Done :)

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (PhilippeMoussalli) [reply examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb]: I Included a few also in the reference for general guidelines on data transformation

**Comment by KevinGG** [comment]
@PhilippeMoussalli Could you please take a look at https://github.com/apache/beam/pull/23069?

**Comment by PhilippeMoussalli** [comment]
> @PhilippeMoussalli Could you please take a look at #23069?

@KevinGG I just tested it out and it checks out! Thanks again for taking this up.

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (TheNeuralBit) [reply examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:1617]: Hm that's odd. This could be a bug. @yeandy do we have a way to apply a categorical DType to a columnn?

Comment (TheNeuralBit) [reply examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:1617]: Interesting apparently instances of CategoricalDtype are considered equal to the string "category" : https://github.com/pandas-dev/pandas/blob/54347fe684e0f7844bf407b1fb958a5269646825/pandas/core/dtypes/dtypes.py#L366

The aim in our check was to avoid the case where users indicate `astype("category")` and rely on pandas to resolve the categories, since we need explicit categories. We should be able to find another way to check this, but it will have to be another bugfix.

Comment (yeandy) [reply examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:1617]: [Here](https://github.com/apache/beam/blob/master/sdks/python/apache_beam/dataframe/frames_test.py#L2417) and [here](https://github.com/apache/beam/blob/master/sdks/python/apache_beam/dataframe/frames_test.py#L1486) are some examples of applying `CategoricalDtype`

Comment (TheNeuralBit) [reply examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:1617]: Ah but both of those examples apply the categorical type with astype on a pandas Series, outside of the test framework. It seems we don't verify it with DeferredSeries.astype

Comment (TheNeuralBit) [reply examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:1617]: Filed #23276 to track this bug.

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (davidcavazos) [reply examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:1979]: Is this where `source_csv_file` should go?

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (davidcavazos) [inline examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:2017]: Nit: can we default to `us-central1`? Its Carbon Free Energy percentage is higher and might have a more average latency for the east coast.

https://cloud.google.com/sustainability/region-carbon

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (davidcavazos) [inline examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:2122]: Feature columns are no longer recommended. They were used for tf.Estimators, but now that TF2 has moved to Keras they now recommend the preprocessing layers.

The preprocessing layers already take care of normalizing with the z-score and one hot encodings, but it's still nice seeing how to do that with Beam.

**Review by davidcavazos — COMMENTED** [review]
Thanks! This looks great! I love the asteroids data used for this sample.

It would be nice to include a small paragraph saying what we're trying to do, like "Creating a preprocessed dataset for a machine learning model in Beam DataFrames". I wasn't sure what the goal of the notebook was until almost at the end. It would also be nice to mention it only creates the dataset, not the model.

I also noticed that the Keras preprocessing layers already take care of normalizing and one-hot-encoding data. I tend to recommend this in my own samples because it allows us to pass the model the raw data like we see it in the field, and let the model itself normalize it as part of its architecture. And adding more data to the dataset does not require normalizing the entire thing again, it could be simply adding more files. However, I think it's nice to see how to do that with Beam DataFrames as well. I think it's particularly important to explore and understand the data.

Other than that, I really liked the way you presented and explored the data in a very interactive way.

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (TheNeuralBit) [reply examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:1866]: The issue with astype should be resolved now and the fix will be in Beam 2.43.0. Can you try this again?

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (PhilippeMoussalli) [reply examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:1421]: woops nice catch ;)

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (PhilippeMoussalli) [reply examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:183]: Sure thing. I'll add it

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (PhilippeMoussalli) [reply examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:2017]: easy fix!

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (PhilippeMoussalli) [reply examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:2122]: good to know, i'll remove the resource

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (PhilippeMoussalli) [reply examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:110]: There was another feature that was recently implemented by @TheNeuralBit to implement the one-hot-encoding and that's scheduled for 2.43. I'll update the comment in that section
https://github.com/apache/beam/issues/23276

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (PhilippeMoussalli) [reply examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:1866]: thanks @TheNeuralBit ! Indeed this fix seems to the job. I'll submit a PR soon with the latest updates

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (PhilippeMoussalli) [reply examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:1979]: > Thanks! This looks great! I love the asteroids data used for this sample.
> 
> It would be nice to include a small paragraph saying what we're trying to do, like "Creating a preprocessed dataset for a machine learning model in Beam DataFrames". I wasn't sure what the goal of the notebook was until almost at the end. It would also be nice to mention it only creates the dataset, not the model.
> 
> I also noticed that the Keras preprocessing layers already take care of normalizing and one-hot-encoding data. I tend to recommend this in my own samples because it allows us to pass the model the raw data like we see it in the field, and let the model itself normalize it as part of its architecture. And adding more data to the dataset does not require normalizing the entire thing again, it could be simply adding more files. However, I think it's nice to see how to do that with Beam DataFrames as well. I think it's particularly important to explore and understand the data.
> 
> Other than that, I really liked the way you presented and explored the data in a very interactive way.

Thanks for the feedback @davidcavazos :). Indeed the preprocessing steps can be done with Keras. I think the ebenfit of showing how it can be done in Pandas is to allow for the possibility to train with other frameworks (sklearn, XGboost,...)

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (PhilippeMoussalli) [reply examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:183]: I just read more about the `splittable` param and it turns out it's not needed for this example. i'll leave it out

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (PhilippeMoussalli) [reply examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:2122]: Good to know. I'll remove the reference

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (TheNeuralBit) [inline examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:96]: Could we file an issue to track removing this once Beam 2.43.0 is released?

Comment (TheNeuralBit) [reply examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:96]: It would be nice to add a TODO in here as well, but I wouldn't want it to be visible to the user. I'm not sure if there's a way to do that in the ipynb format.

**Review by TheNeuralBit — APPROVED** [review]
LGTM, thanks so much @PhilippeMoussalli, this is awesome and I really appreciate your patience :)

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (davidcavazos) [reply examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:1529]: It looks like this typo is still here :)

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (davidcavazos) [inline examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:1885]: one-hot encod*ing* variables -> one-hot encod*ed* variables (?)

**Review by davidcavazos — APPROVED** [review]
A couple tiny typos. LGTM! Thank you, this is great! 🙂

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (TheNeuralBit) [reply examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:96]: It looks like, in theory, we could stick this in a custom [metadata field](https://nbformat.readthedocs.io/en/latest/format_description.html#metadata), something like:
```
      "metadata": {
        "id": "-OJC0Xn5Um-C",
        "beam:comment": "TODO(https://github.com/apache/beam/XXXX): Just install 2.43.0 once it's released",
      },
```

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (davidcavazos) [inline examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:1529]: Typo: trainign -> training (still present)

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (PhilippeMoussalli) [reply examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:97]: done!

**Comment by PhilippeMoussalli** [comment]
Implemented latest feedback @TheNeuralBit @davidcavazos :)

**Comment by damccorm** [comment]
Run Website PreCommit

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (damccorm) [inline examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:94]:
```suggestion
        "beam:comment": "TODO(https://github.com/apache/beam/issues/23961): Just install 2.43.0 once it's released, [`issue 23276`](https://github.com/apache/beam/issues/23276)  is currently not implemented for Beam 2.42 (required fix for implementing `str.get_dummies()`"
```

I filed an issue for this, could we reference it directly?

File: examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb
Comment (davidcavazos) [reply examples/notebooks/beam-ml/dataframe_api_preprocessing.ipynb:97]: I don't think JSON accepts comments, but it could be an `<!-- HTML comment -->` in a Markdown cell.

**Pull request merged**
